File: 0213-house-robber-ii/0213-house-robber-ii.py
from typing import List

class Solution:
    def f(self, ind: int, num: List[int], dp: List[int]) -> int:
        if ind < 0:
            return 0
        if ind == 0:
            return num[0]
        if dp[ind] != -1:
            return dp[ind]
        
        # Two choices: don't take the current house, or take it and skip the previous one
        not_take = self.f(ind - 1, num, dp)
        take = num[ind] + self.f(ind - 2, num, dp)
        dp[ind] = max(take, not_take)
        return dp[ind]

    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        
        # Case 1: Rob houses from 0 to n-2 (skip the last house)
        temp1 = nums[:-1]
        dp1 = [-1] * len(temp1)
        first = self.f(len(temp1) - 1, temp1, dp1)
        
        # Case 2: Rob houses from 1 to n-1 (skip the first house)
        temp2 = nums[1:]
        dp2 = [-1] * len(temp2)
        second = self.f(len(temp2) - 1, temp2, dp2)
        
        # The maximum of both cases gives the result
        return max(first, second)

        

        

File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
    private:
    int f(int ind,vector<int>& num,  vector<int>& dp){
        if(ind<0) return 0;
        if(ind ==0) return num[0];
        if(dp[ind]!=-1) return dp[ind];


        int notTake = 0+f(ind-1,num,dp);
        int take = num[ind]+f(ind-2,num,dp);
        return dp[ind]=max(take,notTake);
    }

public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n,-1);
        if(n==1) return nums[0];
         vector<int> temp1(begin(nums), end(nums)-1);
        vector<int> temp2(begin(nums)+1, end(nums));
        int size = temp1.size();
        int first = f(size-1,temp1,dp);
        dp.clear();
        dp.assign(n, -1);
        int second = f(size-1,temp2,dp);
        return max(first,second);
       
        
    }
};